rust-connectors/sources/postgres/src/connect.rs

```diff
@@ -0,0 +1,170 @@
+use crate::model::{Column, LogicalReplicationMessage, ReplicationEvent};
+use crate::{Error, PgConnectorOpt};
+use fluvio::metadata::topic::TopicSpec;
+use fluvio::{Fluvio, Offset, TopicProducer};
+use once_cell::sync::Lazy;
+use postgres_protocol::message::backend::{
+    LogicalReplicationMessage as PgReplication, ReplicationMessage,
+};
+use std::collections::BTreeMap;
+use std::pin::Pin;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+use tokio_postgres::config::ReplicationMode;
+use tokio_postgres::replication::LogicalReplicationStream;
+use tokio_postgres::types::PgLsn;
+use tokio_postgres::{Client, NoTls};
+use tokio_stream::StreamExt;
+
+const TIME_SEC_CONVERSION: u64 = 946_684_800;
+static EPOCH: Lazy<SystemTime> =
+    Lazy::new(|| UNIX_EPOCH + Duration::from_secs(TIME_SEC_CONVERSION));
+
+/// A Fluvio connector for Postgres CDC.
+pub struct PgConnector {
+    /// The connector configuration.
+    config: PgConnectorOpt,
+    /// The Postgres client for streaming replication changes.
+    pg_client: Client,
+    /// The Fluvio producer for recording change events.
+    producer: TopicProducer,
+    /// The current Log Sequence Number (offset) to read in the replication stream.
+    lsn: Option<PgLsn>,
+    /// Caches the schema for each new table we see, grouped by relation_id
+    relations: BTreeMap<u32, Vec<Column>>,
+}
+
+impl PgConnector {
+    pub async fn new(config: PgConnectorOpt) -> eyre::Result<Self> {
+        tracing::info!("Initializing PgConnector");
+        let fluvio = Fluvio::connect().await?;
+        tracing::info!("Connected to Fluvio");
+
+        let admin = fluvio.admin().await;
+        let topics = admin.list::<TopicSpec, _>(vec![]).await?;
+        let topic_exists = topics.iter().any(|t| t.name == config.topic);
+        if !topic_exists {
+            return Err(Error::TopicNotFound(config.topic.to_string()).into());
+        }
+
+        let consumer = fluvio.partition_consumer(&config.topic, 0).await?;
+        let mut lsn: Option<PgLsn> = None;
+
+        // Try to get the last item from the Fluvio Topic. Timeout after 1 second
+        let stream = consumer.stream(Offset::from_end(1)).await?;
+        let timeout = stream.timeout(Duration::from_millis(1_000));
```

Comment:
To resume the Postgres stream, we need to know the most recently seen Postgres LSN (offset), which would be in the last Fluvio record. However, when fetching latest record, it will simply hang if there are no records at all, so I put a timeout, after which we'll treat it as if there is no records (i.e start from beginning).

I can put this into the configuration if you like 👍

Reply:
Added as `resume-timeout`